cacti/tests/test_cacti.py

```diff
@@ -0,0 +1,171 @@
+    'rra[6].cf': 'MAX',
+    'rra[7].cdp_prep[0].unknown_datapoints': 177L,
+    'rra[1].cdp_prep[0].unknown_datapoints': 0L,
+    'rra[2].rows': 775L,
+    'rra[2].pdp_per_row': 24L,
+}
+
+ts = int(time.time())
+MOCK_FETCH = ((ts - 300, ts, 300), ('mem_buffers',), [(2048,), (None,)])
+
+MOCK_RRD_META = [
+    ('localhost', None, '/var/lib/cacti/rra/localhost_mem_buffers_3.rrd'),
+    ('localhost', None, '/var/lib/cacti/rra/localhost_mem_swap_4.rrd'),
+    ('localhost', None, '/var/lib/cacti/rra/localhost_load_1min_5.rrd'),
+    ('localhost', None, '/var/lib/cacti/rra/localhost_users_6.rrd'),
+    ('localhost', None, '/var/lib/cacti/rra/localhost_proc_7.rrd'),
+]
+
+
+CHECK_NAME = 'cacti'
+
+CUSTOM_TAGS = ['optional:tag1']
+
+CACTI_CONFIG = {
+    'mysql_host': 'nohost',
+    'mysql_user': 'mocked',
+    'rrd_path': '/rrdtool/is/mocked',
+    'tags': CUSTOM_TAGS,
+}
+
+
+@pytest.fixture
+def aggregator():
+    from datadog_checks.stubs import aggregator
+    aggregator.reset()
+    return aggregator
+
+
+@pytest.fixture
+def check():
+    return Cacti(CHECK_NAME, {}, {})
+
+
+def test_check(aggregator, check):
+
+    mocks = [
+        mock.patch('datadog_checks.cacti.cacti.pymysql'),
+        mock.patch('datadog_checks.cacti.cacti.rrdtool'),
+        mock.patch('datadog_checks.cacti.Cacti._get_rrd_info', return_value=MOCK_INFO),
+        mock.patch('datadog_checks.cacti.Cacti._get_rrd_fetch', return_value=MOCK_FETCH),
+        mock.patch('datadog_checks.cacti.Cacti._fetch_rrd_meta', return_value=MOCK_RRD_META),
+    ]
+
+    for mock_func in mocks:
+        mock_func.start()
+
+    # Run the check twice to set the timestamps and capture metrics on the second run
+    check.check(CACTI_CONFIG)
+    check.check(CACTI_CONFIG)
+
+    for mock_func in mocks:
+        mock_func.stop()
+
+    # We are mocking the MySQL call so we won't have cacti.rrd.count or cacti.hosts.count metrics,
+    # check for metrics that are returned from our mock data.
+    aggregator.assert_metric('cacti.metrics.count', value=10, tags=CUSTOM_TAGS)
+    aggregator.assert_metric('system.mem.buffered.max', value=2, tags=CUSTOM_TAGS)
+    aggregator.assert_metric('system.mem.buffered', value=2, tags=CUSTOM_TAGS)
+
+
+def test_cacti_metrics(aggregator, check):
+
+    mocks = [
+        mock.patch('datadog_checks.cacti.cacti.pymysql'),
+        mock.patch('datadog_checks.cacti.cacti.rrdtool'),
+    ]
+
+    for mock_func in mocks:
+        mock_func.start()
+
+    # Run the check twice to set the timestamps and capture metrics on the second run
+    check.check(CACTI_CONFIG)
+    check.check(CACTI_CONFIG)
+
+    for mock_func in mocks:
+        mock_func.stop()
+
+    # We are mocking pymysql, this results in returning only the cacti.* metrics.
+    aggregator.assert_metric('cacti.metrics.count', value=0, tags=CUSTOM_TAGS)
+    aggregator.assert_metric('cacti.rrd.count', value=0, tags=CUSTOM_TAGS)
+    aggregator.assert_metric('cacti.hosts.count', value=0, tags=CUSTOM_TAGS)
```

Comment:
This is gonna take some work but can we actually mock ``pymysql.connect.cursor.execute``? I'd love to see the actual SQL queries be mocked.

That way, these metric values won't be 0 👍